ddl/ddl_api.go

```diff
@@ -3372,175 +3372,135 @@ func checkModifyCharsetAndCollation(toCharset, toCollate, origCharset, origColla
 
 // CheckModifyTypeCompatible checks whether changes column type to another is compatible considering
 // field length and precision.
-func CheckModifyTypeCompatible(origin *types.FieldType, to *types.FieldType) (allowedChangeColumnValueMsg string, err error) {
-	unsupportedMsg := fmt.Sprintf("type %v not match origin %v", to.CompactStr(), origin.CompactStr())
-	var skipSignCheck bool
-	var skipLenCheck bool
-	switch origin.Tp {
-	case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString, mysql.TypeBlob,
-		mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
-		switch to.Tp {
-		case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString,
-			mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
-			skipSignCheck = true
-			skipLenCheck = true
-		case mysql.TypeBit:
-			// TODO: Currently string data type cast to bit are not compatible with mysql, should fix here after compatible.
-			return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		default:
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+func CheckModifyTypeCompatible(origin *types.FieldType, to *types.FieldType) (canReorg bool, errMsg string, err error) {
+	checkLengthDecimalSign := func() (bool, string, error) {
```

Comment:
I prefer to add a new function instead of a function pointer

Reply:
My point is limit the logic into smaller scope since this function is only used inside the CheckModifyTypeCompatible() and the return value and error type are also special purpose.